File: Caso2/TypeScript/Back/main.ts
/**
 * Runner de CentralAscensores: lee un archivo de comandos y ejecuta cada
 * linea, imprimiendo el resultado en el mismo estilo del PDF.
 *
 * Uso: npx ts-node main.ts central.txt
 *
 * Gramatica del archivo (una instruccion por linea, '#' inicia comentario):
 *
 *   PARQUE
 *   <codigo>, <edificio>, <estado>, <emergencias>
 *   ...
 *   ---
 *   LLAMADA <codigo> <EMERGENCIA|MANTENIMIENTO> <minuto>
 *   ATENDER <minutoAtencion>
 *   PASO <n>
 *   DESHACER
 *   ABORTAR <motivo>
 *   CERRAR
 *   REPORTE
 */
import * as fs from "fs";
import { CentralAscensores } from "./central";
import { RechazoOperacion, TipoLlamada } from "./entidades";

function quitarComentario(linea: string): string {
  return linea.split("#")[0].trim();
}

function cargarArchivo(ruta: string): { filasParque: string[][]; comandos: string[] } {
  const crudo = fs.readFileSync(ruta, "utf-8").split(/\r?\n/);
  const lineas = crudo.map(quitarComentario).filter((l) => l.length > 0);

  if (lineas[0].toUpperCase() !== "PARQUE") {
    throw new Error("el archivo debe empezar con PARQUE");
  }
  const filasParque: string[][] = [];
  let i = 1;
  while (lineas[i] !== "---") {
    filasParque.push(lineas[i].split(",").map((p) => p.trim()));
    i++;
  }
  const comandos = lineas.slice(i + 1);
  return { filasParque, comandos };
}

function ejecutar(ruta: string): void {
  const { filasParque, comandos } = cargarArchivo(ruta);
  const central = new CentralAscensores();
  central.cargarParque(filasParque);

  for (const linea of comandos) {
    console.log(`> ${linea}`);
    const partes = linea.split(/\s+/);
    const cmd = partes[0].toUpperCase();
    try {
      switch (cmd) {
        case "LLAMADA": {
          const [, codigo, tipo, minuto] = partes;
          console.log(" ", central.registrarLlamada(codigo, tipo as TipoLlamada, Number(minuto)));
          break;
        }
        case "ATENDER":
          console.log(" ", central.atenderSiguiente(Number(partes[1])));
          break;
        case "PASO":
          console.log(" ", central.ejecutarPaso(Number(partes[1])));
          break;
        case "DESHACER":
          console.log(" ", central.deshacerUltimo());
          break;
        case "ABORTAR":
          for (const l of central.abortarRescate(partes.slice(1).join(" "))) {
            console.log(" ", l);
          }
          break;
        case "CERRAR":
          console.log(" ", central.cerrarLlamada());
          break;
        case "REPORTE":
          console.log(central.reporte());
          break;
        default:
          console.log(`  comando desconocido: ${cmd}`);
      }
    } catch (err) {
      if (err instanceof RechazoOperacion) {
        console.log(`  RECHAZADA ${err.message}`);
      } else {
        throw err;
      }
    }
  }
}

const ruta = process.argv[2] ?? "central.txt";
ejecutar(ruta);

File: Caso2/Python/Back/main.py
"""
Runner de CentralAscensores: lee un archivo de comandos y ejecuta cada
linea sobre una instancia de CentralAscensores, imprimiendo el resultado
en el mismo estilo del PDF.

Uso:
    python main.py central.txt

Gramatica del archivo (una instruccion por linea, '#' inicia comentario):

    PARQUE
    <codigo>, <edificio>, <estado>, <emergencias>
    ...
    ---
    LLAMADA <codigo> <EMERGENCIA|MANTENIMIENTO> <minuto>
    ATENDER <minutoAtencion>
    PASO <n>
    DESHACER
    ABORTAR <motivo>
    CERRAR
    REPORTE
"""

import sys

from central import CentralAscensores
from entidades import RechazoOperacion


def quitar_comentario(linea: str) -> str:
    return linea.split("#", 1)[0].strip()


def cargar_archivo(ruta: str):
    with open(ruta, encoding="utf-8") as f:
        crudo = [linea.rstrip("\n") for linea in f]

    lineas = [quitar_comentario(l) for l in crudo]
    lineas = [l for l in lineas if l]

    assert lineas[0].upper() == "PARQUE", "el archivo debe empezar con PARQUE"
    filas_parque = []
    i = 1
    while lineas[i] != "---":
        partes = [p.strip() for p in lineas[i].split(",")]
        filas_parque.append(partes)
        i += 1
    comandos = lineas[i + 1:]
    return filas_parque, comandos


def procesar_linea(central: CentralAscensores, linea: str) -> list[str]:
    """Ejecuta una linea de comando sobre `central` y devuelve el resultado
    como lista de lineas (una sola para la mayoria de comandos, varias para
    ABORTAR y REPORTE). No imprime nada: la usan tanto main.py como el Front
    para no duplicar el despacho de comandos en dos lugares.
    """
    partes = linea.split()
    cmd = partes[0].upper()
    try:
        if cmd == "LLAMADA":
            _, codigo, tipo, minuto = partes
            return [central.registrar_llamada(codigo, tipo, int(minuto))]
        elif cmd == "ATENDER":
            return [central.atender_siguiente(int(partes[1]))]
        elif cmd == "PASO":
            return [central.ejecutar_paso(int(partes[1]))]
        elif cmd == "DESHACER":
            return [central.deshacer_ultimo()]
        elif cmd == "ABORTAR":
            return central.abortar_rescate(" ".join(partes[1:]))
        elif cmd == "CERRAR":
            return [central.cerrar_llamada()]
        elif cmd == "REPORTE":
            return central.reporte().splitlines()
        else:
            return [f"comando desconocido: {cmd}"]
    except RechazoOperacion as err:
        return [f"RECHAZADA {err}"]


def ejecutar(ruta: str) -> None:
    filas_parque, comandos = cargar_archivo(ruta)
    central = CentralAscensores()
    central.cargar_parque(filas_parque)

    for linea in comandos:
        print(f"> {linea}")
        cmd = linea.split()[0].upper()
        resultado = procesar_linea(central, linea)
        if cmd == "REPORTE":
            print("\n".join(resultado))
        else:
            for l in resultado:
                print(" ", l)


if __name__ == "__main__":
    ruta = sys.argv[1] if len(sys.argv) > 1 else "central.txt"
    ejecutar(ruta)
